File: test/trace_processor/diff_tests/stdlib/traced/stats.py
#!/usr/bin/env python3
# Copyright (C) 2025 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from python.generators.diff_tests.testing import Path, DataPath, Metric
from python.generators.diff_tests.testing import Csv, Json, TextProto
from python.generators.diff_tests.testing import DiffTestBlueprint
from python.generators.diff_tests.testing import TestSuite

BASE_TRACE = r"""
        packet {
          timestamp: 500000
          trace_uuid {
            msb: 1
            lsb: 2
          }
        }"""

CLONE_TRIGGERED = r"""
        packet {
          timestamp: 500000
          clone_snapshot_trigger {
            trigger_name: "trigger_name"
          }
        }
        """

CLONE_STARTED = r"""
        packet {
          timestamp: 1000000
          service_event {
            clone_started: true
          }
        }
        """

CLONE_DONE = r"""
        packet {
          timestamp: 2000000
          service_event {
            buffer_cloned: 1
          }
        }
        packet {
          timestamp: 3000000
          service_event {
            buffer_cloned: 2
          }
        }
        packet {
          timestamp: 4000000
          service_event {
            buffer_cloned: 0
          }
        }
        """


class TracedStats(TestSuite):

  def test_clone_flush_latency(self):
    return DiffTestBlueprint(
        trace=TextProto(BASE_TRACE + CLONE_STARTED + CLONE_DONE),
        query=r"""
        INCLUDE PERFETTO MODULE traced.stats;

        SELECT * FROM traced_clone_flush_latency;
        """,
        out=Csv("""
        "buffer_id","duration_ns"
        0,3000000
        1,1000000
        2,2000000
        """))

  def test_clone_flush_latency_missing_clone_started(self):
    return DiffTestBlueprint(
        trace=TextProto(BASE_TRACE + CLONE_DONE),
        query=r"""
        INCLUDE PERFETTO MODULE traced.stats;

        SELECT * FROM traced_clone_flush_latency;
        """,
        out=Csv("""
        "buffer_id","duration_ns"
        """))

  def test_clone_flush_latency_missing_buffer_cloned(self):
    return DiffTestBlueprint(
        trace=TextProto(BASE_TRACE + CLONE_STARTED),
        query=r"""
        INCLUDE PERFETTO MODULE traced.stats;

        SELECT * FROM traced_clone_flush_latency;
        """,
        out=Csv("""
        "buffer_id","duration_ns"
        """))

  def test_trigger_clone_flush_latency(self):
    return DiffTestBlueprint(
        trace=TextProto(BASE_TRACE + CLONE_TRIGGERED + CLONE_DONE),
        query=r"""
        INCLUDE PERFETTO MODULE traced.stats;

        SELECT * FROM traced_trigger_clone_flush_latency;
        """,
        out=Csv("""
        "buffer_id","duration_ns"
        0,3500000
        1,1500000
        2,2500000
        """))

  def test_trigger_clone_flush_latency_missing_clone_started(self):
    return DiffTestBlueprint(
        trace=TextProto(BASE_TRACE + CLONE_DONE),
        query=r"""
        INCLUDE PERFETTO MODULE traced.stats;

        SELECT * FROM traced_trigger_clone_flush_latency;
        """,
        out=Csv("""
        "buffer_id","duration_ns"
        """))

  def test_trigger_clone_flush_latency_missing_buffer_cloned(self):
    return DiffTestBlueprint(
        trace=TextProto(BASE_TRACE + CLONE_TRIGGERED),
        query=r"""
        INCLUDE PERFETTO MODULE traced.stats;

        SELECT * FROM traced_trigger_clone_flush_latency;
        """,
        out=Csv("""
        "buffer_id","duration_ns"
        """))

File: ui/src/plugins/dev.perfetto.TraceInfoPage/tabs/stats.ts
// Copyright (C) 2025 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import m from 'mithril';
import type {Engine} from '../../../trace_processor/engine';
import {NUM_NULL} from '../../../trace_processor/query_result';
import {Button} from '../../../widgets/button';
import {Section} from '../../../widgets/section';
import {Grid, GridCell, GridHeaderCell} from '../../../widgets/grid';
import {Icon} from '../../../widgets/icon';
import {Tooltip} from '../../../widgets/tooltip';
import {statsSpec, type StatsSectionRow} from '../utils';

export interface StatsData {
  allStats: StatsSectionRow[];
  isMultiTrace: boolean;
  isMultiMachine: boolean;
}

export async function loadStatsData(engine: Engine): Promise<StatsData> {
  // Load all stats
  const allStatsResult = await engine.query(`
    select
      name,
      value,
      cast(ifnull(idx, '') as text) as idx,
      description,
      severity,
      source,
      machine_id as machineId,
      trace_id as traceId
    from stats
    order by trace_id, machine_id, name, idx
  `);
  const allStats: StatsSectionRow[] = [];
  for (const iter = allStatsResult.iter(statsSpec); iter.valid(); iter.next()) {
    allStats.push({
      name: iter.name,
      value: iter.value,
      description: iter.description,
      idx: iter.idx,
      severity: iter.severity,
      source: iter.source,
      machineId: iter.machineId,
      traceId: iter.traceId,
    });
  }

  // Determine whether the trace spans multiple traces/machines so we know
  // whether to show the per-trace/per-machine columns. The metadata table is
  // the authoritative source of trace/machine ids; supplement it with any ids
  // observed only in the stats rows.
  const traceIds = new Set<number>();
  const machineIds = new Set<number>();

  const idResult = await engine.query(`
    select distinct trace_id as traceId, machine_id as machineId
    from metadata
  `);
  for (
    const iter = idResult.iter({traceId: NUM_NULL, machineId: NUM_NULL});
    iter.valid();
    iter.next()
  ) {
    if (iter.traceId !== null) {
      traceIds.add(iter.traceId);
    }
    if (iter.machineId !== null) {
      machineIds.add(iter.machineId);
    }
  }
  for (const stat of allStats) {
    if (stat.machineId !== null) {
      machineIds.add(stat.machineId);
    }
    if (stat.traceId !== null) {
      traceIds.add(stat.traceId);
    }
  }

  return {
    allStats,
    isMultiTrace: traceIds.size > 1,
    isMultiMachine: machineIds.size > 1,
  };
}

export interface StatsTabAttrs {
  data: StatsData;
}

export class StatsTab implements m.ClassComponent<StatsTabAttrs> {
  view({attrs}: m.CVnode<StatsTabAttrs>) {
    return m(
      '.pf-trace-info-page__tab-content',
      m(
        Section,
        {
          title: 'Statistics',
          subtitle:
            'Complete dump of all trace statistics including errors, data losses, and debugging info',
        },
        m(StatsSection, {
          data: attrs.data.allStats,
          isMultiTrace: attrs.data.isMultiTrace,
          isMultiMachine: attrs.data.isMultiMachine,
        }),
      ),
    );
  }
}

// Stats Section
interface StatsSectionAttrs {
  data: StatsSectionRow[];
  isMultiTrace: boolean;
  isMultiMachine: boolean;
}

class StatsSection implements m.ClassComponent<StatsSectionAttrs> {
  private hideZeroValues = true;

  view({attrs}: m.CVnode<StatsSectionAttrs>) {
    const data = attrs.data;
    if (data === undefined || data.length === 0) {
      return m('');
    }

    const filtered = this.hideZeroValues
      ? data.filter((row) => row.value !== 0 && row.value !== null)
      : data;

    const columns = [
      ...(attrs.isMultiTrace
        ? [{key: 'trace', header: m(GridHeaderCell, 'Trace')}]
        : []),
      ...(attrs.isMultiMachine
        ? [{key: 'machine', header: m(GridHeaderCell, 'Machine')}]
        : []),
      {key: 'name', header: m(GridHeaderCell, 'Name')},
      {key: 'value', header: m(GridHeaderCell, 'Value')},
      {key: 'type', header: m(GridHeaderCell, 'Type')},
    ];

    const rowData = filtered.map((row) => {
      const idx = row.idx !== '' ? `[${row.idx}]` : '';
      const help = Boolean(row.description)
        ? m(
            Tooltip,
            {
              trigger: m(Icon, {
                icon: 'help_outline',
                className: 'pf-trace-info-page__help-icon',
              }),
            },
            `${row.description}`,
          )
        : undefined;
      const cells = [];
      if (attrs.isMultiTrace) {
        cells.push(m(GridCell, row.traceId === null ? '-' : row.traceId));
      }
      if (attrs.isMultiMachine) {
        cells.push(m(GridCell, row.machineId === null ? '-' : row.machineId));
      }
      cells.push(
        m(
          GridCell,
          {className: 'pf-trace-info-page__grid-key'},
          `${row.name}${idx}`,
          help,
        ),
        m(GridCell, `${row.value}`),
        m(GridCell, `${row.severity} (${row.source})`),
      );
      return cells;
    });

    return m(
      'section.pf-trace-info-page__stats-section',
      m(Button, {
        label: this.hideZeroValues ? 'Show zero values' : 'Hide zero values',
        icon: this.hideZeroValues ? 'visibility' : 'visibility_off',
        onclick: () => {
          this.hideZeroValues = !this.hideZeroValues;
        },
      }),
      m(Grid, {
        columns,
        rowData,
        className: 'pf-trace-info-page__dense-grid',
      }),
    );
  }
}
